src/Analyser/NodeScopeResolver.php

```diff
@@ -4291,6 +4292,24 @@ private function processClosureNode(
 			return new ProcessClosureResult($scope, $statementResult->getThrowPoints(), $statementResult->getImpurePoints(), $invalidateExpressions);
 		}
 
+		if ($expr->getAttribute(ImmediatelyInvokedClosureVisitor::ATTRIBUTE_NAME) === true) {
+			$intermediaryClosureScopeResult = $this->processStmtNodes($expr, $expr->stmts, $closureScope, static function (): void {
+			}, StatementContext::createTopLevel());
+			$intermediaryClosureScope = $intermediaryClosureScopeResult->getScope();
+
+			$statementResult = $this->processStmtNodes($expr, $expr->stmts, $closureScope, $closureStmtsCallback, StatementContext::createTopLevel());
+			$nodeCallback(new ClosureReturnStatementsNode(
+				$expr,
+				$gatheredReturnStatements,
+				$gatheredYieldStatements,
+				$statementResult,
+				$executionEnds,
+				array_merge($statementResult->getImpurePoints(), $closureImpurePoints),
+			), $closureScope);
+
+			return new ProcessClosureResult($scope->processClosureScope($intermediaryClosureScope, null, $byRefUses), $statementResult->getThrowPoints(), $statementResult->getImpurePoints(), $invalidateExpressions);
```

Comment:
in case we agree this fix is correct, I think we should add new tests for IIFE regarding impure-points and throw-points